full-moon/src/atom.rs

```diff
@@ -0,0 +1,291 @@
+use logos::{Lexer, Logos};
+
+use crate::ShortString;
+
+pub(crate) fn trim_bracket_head(slice: &str) -> (ShortString, Option<usize>) {
+    match test_bracket_head(slice) {
+        Some(count) => {
+            let trim = &slice[count + 2..slice.len() - count - 2];
+
+            (trim.into(), Some(count))
+        }
+        None => (slice.into(), None),
+    }
+}
+
+fn test_bracket_head(slice: &str) -> Option<usize> {
+    // starts with `[`?
+    if !slice.starts_with('[') {
+        return None;
+    }
+
+    // how many `=` after `[`?
+    let count = slice.chars().skip(1).take_while(|&v| v == '=').count();
+
+    // ends with `[`?
+    if !matches!(slice.chars().nth(count + 1), Some('[')) {
+        return None;
+    }
+
+    Some(count)
+}
+
+fn read_bracketed(lex: &mut Lexer<Atom>, skips: usize) -> bool {
+    let num_eq = match lex.slice().get(skips..).and_then(test_bracket_head) {
```

Comment:
Instead of using `and_then(test_bracket_head)`, just run `test_bracket_head(value)` inside the `Some` branch.

Reply:
In this case `test_bracket_head` returns an `Option`, so running it with `and_then` ensures both the slice is valid and that it has a proper head.